proposals/0000-extension-lifecycle-framework.md

```diff
@@ -0,0 +1,128 @@
+
+ * `Deprecated` extensions are considered to be design or implementation dead ends, and should probably not be used in new code. Deprecating an extension _must_ come with a statement about the longevity of the extension. This need not be a commitment to remove it at a concrete time, but is a commitment to either remove it or transition it to another state at a time stated by GHC HQ when it is marked as `Deprecated`. Any extension will be deprecated prior to removal. Deprecated extensions are expected to include `IncoherentInstances`, `OverlappingInstances`, `Rank2Types`, `RecordPuns`, and `TypeInType`.
+ 
+ * `Legacy` extensions are explicitly not recommended for new code, much like `Deprecated`, however, they are expected to be supported indefinitely. This may be to maintain an older language set, maintain a bridge to a newer feature, or otherwise keep backward compatibility. Expected legacy extensions include `CUSKs`, `DeepSubsumption`,  and `NPlusKPatterns`.
+
+Above we list expected categorizations of several extensions based on statements in the user's guide. We do not attempt to prescribe any particular categorizations as part of this proposal and simply provide a small number of expectations from the perspective of someone reading the existing documentation.
+
+The expected extension lifecycle includes the following transitions:
+ * `Experimental` -> `Mature`
+ * `Experimental` -> `Deprecated`
+ * `Mature` -> `Legacy`
+ * `Mature` -> `Deprecated`
+ * `Legacy` -> `Deprecated`
+ * `Deprecated` -> `Legacy`
+
+However, it also seems plausible that new knowledge might from time to time cause a mature extension to once again be considered experimental, e.g. in the face of soundness bugs or subtle interactions with other features. We also do not rule out any transition explicitly to allow for unforeseen circumstances. 
+
+For existing, or future, language sets such as `GHC2021` or `Haskell98`, it is expected that none of the contained extensions would be `Experimental`. However, this proposal does not seek to impose any particular policy on the inclusion of extensions into language sets - the developers and the steering committee are always in the best position to make a decision about a concrete extension and extension set.
+
+### 2.2. User Interface
+
+Haskell users have different tolerances for risks related to language change. Some derive great value from using the newest features, while others are conservative. GHC should serve the whole spectrum.
+
+There will be an additional set of warnings:
+ * `-WXDeprecated`: Issue a warning when a deprecated extension is used
```

Comment:
Please can the proposal include a definition of what it means for an extension to be "used"? I suspect it means "when a `LANGUAGE` pragma in a module enables the extension", which is easily specified, and wouldn't warn if the extension was enabled in a `.cabal` file or on the command line. But it could also mean "when the feature enabled by the extension occurs in the program" (e.g. an occurrence of a string literal for `OverloadedStrings`).

Reply:
I'd expected this to mean ”enabled”, independently of how (pragma, flag, cabal file).

I'd advise against differtiating between these ways to enable the extension, it would complicate the conceptual meta-theory of extensions. (Also, I am not sure if GHC even knows whether a flag was passed by the user, or by `cabal`).

```suggestion
 * `-WXDeprecated`: Issue a warning when a deprecated extension is enabled (by way of pragma, command like flag or otherwise)
```

Reply:
Hmm, but then we have warnings that can arise just based on command-line options, entirely independently of source files. And that seems likely to lead to complications, because command-line options can also modify whether the warnings should be emitted, so it matters exactly when and where the check is done (cf. https://gitlab.haskell.org/ghc/ghc/-/issues/23402).

I can see that e.g. having `ghci -XSomeDeprecatedExtension` issue a warning makes sense, so I'm not strongly opposed to your suggestion. But the proposal should at least specify whether `ghc -WXDeprecated -XSomeDeprecatedExtension -Wno-XDeprecated` does or does not issue a warning.

Reply:
Right, we should be clear about this.

My preference is to keep language pragmas and command line flags equivalent. This means that `-W` flags and `-X` flags commute, and `-X` flags ans pragmas have their effect when actually looking at the file.


Surely, we don’t want developers to think about having to write
```
{-# LANGUAGE SomeDeprecatedExtension #-}
{-# OPTIONS_GHC -Wno-XSomeDeprecatedExtension #-}
```
(which I expect will arise as a common idiom from all this) or
```
{-# OPTIONS_GHC -Wno-XSomeDeprecatedExtension #-}
{-# LANGUAGE SomeDeprecatedExtension #-}
```

I agree that there is existing complexity with warnings about flags, so there would be precedent. Maybe we can keep it contained to that dark corner :-)